### PR TITLE
feat(wing): add parser

### PR DESCRIPTION
**Pull request by gshpychka**
Add a parser for [Wing](https://github.com/winglang/wing)

I've copied the queries from the upstream repo. This is my first contribution here, let me know if I got it right.

### CODE REVIEW

**Comment by clason** [comment]
Thank you! Unfortunately, upstream queries cannot be used (unchanged); the two main differences are
1. we use different capture names, see https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md
2. the query order precedence is different: upstream has "first match wins", Neovim has "all matches are applied, and the last one wins (by painting over the older ones)".

File: lua/nvim-treesitter/parsers.lua [hunk 1]
```diff
@@ -1651,6 +1651,18 @@ list.wgsl_bevy = {
   maintainers = { "@theHamsta" },
 }
 
+list.wing = {
+  install_info = {
+    url = "https://github.com/winglang/wing",
+    branch = "main",
```
[the remark this answers is not in the document]
Comment (gshpychka) [reply lua/nvim-treesitter/parsers.lua:1657]: @clason the docs imply that the default is "master", is that not the case anymore?

[the remark this answers is not in the document]
Comment (clason) [reply lua/nvim-treesitter/parsers.lua:1657]: No, I mean for the parser repo (this is configurable and looks to be `main` for winglang/wing).

**Comment by gshpychka** [comment]
> Thank you! Unfortunately, upstream queries cannot be used (unchanged); the two main differences are
> 
>     1. we use different capture names, see https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md
> 
>     2. the query order precedence is different: upstream has "first match wins", Neovim has "all matches are applied, and the last one wins (by painting over the older ones)".



> Thank you! Unfortunately, upstream queries cannot be used (unchanged); the two main differences are
> 
>     1. we use different capture names, see https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md
> 
>     2. the query order precedence is different: upstream has "first match wins", Neovim has "all matches are applied, and the last one wins (by painting over the older ones)".

I believe I've fixed #1. Regarding #2, can you point me in a direction as to how to fix this? I'm not too familiar with TS.

**Comment by clason** [comment]
Well, best way is to use the parser and see if it behaves as expected; if it doesn't, you can try to just shuffle the individual queries so the ones who should win come _later_ instead of earlier.

(You can also write tests, which are always welcome ;))

File: queries/wing/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  "init"
+  "let"
+  "new"
+  "return"
+  (inflight_specifier)
+] @keyword
+
+(identifier) @variable
+(reference_identifier) @variable
+(member_identifier) @property
```
Comment (clason) [inline queries/wing/highlights.scm:94]: Specifically, I believe these should be at the _top_ of the file here.

**Comment by clason** [comment]
We also provide queries for indentation and folding, by the way. Indentation may be a bit involved, but folds at least should be easy to add (just look at some other ones in this repo).

File: queries/wing/locals.scm [hunk 3]
```diff
@@ -0,0 +1,6 @@
+(block) @local.scope
+(variable_definition_statement
+  name: (identifier) @local.definition
```
Comment (clason) [inline queries/wing/locals.scm:3]: local queries are also different from upstream (see `CONTRIBUTING.md`; we don't use `@local` scope -- for now).

File: lua/nvim-treesitter/parsers.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (gshpychka) [reply lua/nvim-treesitter/parsers.lua:1657]: @clason right, the documentation implies that the default value is "master", and should be overriden if it's something else (like "main"): https://github.com/nvim-treesitter/nvim-treesitter/blob/d49ba798c95398e62321b40041e8ac4654fea251/README.md?plain=1#L505

[the remark this answers is not in the document]
Comment (clason) [reply lua/nvim-treesitter/parsers.lua:1657]: Ah, indeed, that is misleading. You really only need this if the branch is different from the _default_ branch.

(I suspect this was written before the big `master`->`main` migration, when the default branch was fairly universally `master`.)

**Comment by gshpychka** [comment]
> We also provide queries for indentation and folding, by the way. Indentation may be a bit involved, but folds at least should be easy to add (just look at some other ones in this repo).

Added some folds

File: queries/wing/highlights.scm (hunk 2 above)
Comment (gshpychka) [reply queries/wing/highlights.scm]: I committed this change, will need to research further if that alone is not sufficient

File: queries/wing/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+; Classes
+
+(custom_type) @type
+(class_field 
+  name: (identifier) @field
+)
```
Comment (ObserverOfTime) [inline queries/wing/highlights.scm:10]: Lisp formatting convention. Same for the rest.

```suggestion
  name: (identifier) @field)
```

File: queries/wing/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  name: (identifier) @field
+) 
+(class_definition 
+  name: (identifier) @type
+)
+(method_definition
+  name: (identifier) @method
+)
+(inflight_method_definition
+  name: (identifier) @function
```
Comment (ObserverOfTime) [inline queries/wing/highlights.scm:18]: Why is this not `@method`?

File: queries/wing/highlights.scm [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+)
+
+; Primitives
+
+[
+ (number)
+ (duration)
+] @constant.builtin
+(string) @string
+(bool) @constant.builtin
```
Comment (ObserverOfTime) [inline queries/wing/highlights.scm:41]:
```suggestion
(number) @number
(duration) @constant
(string) @string
(bool) @boolean
```

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]

File: queries/wing/highlights.scm [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+(class_field 
+  name: (identifier) @field)
+(class_definition 
+  name: (identifier) @type
+)
+(method_definition
+  name: (identifier) @method
+)
+(inflight_method_definition
+  name: (identifier) @function
```
Comment (gshpychka) [inline queries/wing/highlights.scm:17]:
```suggestion
  name: (identifier) @method
```

File: queries/wing/highlights.scm (hunk 5 above)
Comment (gshpychka) [reply queries/wing/highlights.scm]: This one was my bad. Fixed.

**Comment by gshpychka** [comment]
I am in the process of adding some tests for highlighting

**Comment by gshpychka** [comment]
Added tests, but they fail if I move the last three highlighting queries to the top as suggested [here](https://github.com/nvim-treesitter/nvim-treesitter/pull/5021#discussion_r1242453572)

They pass without this change. Discussing this on matrix, posting here as well for posterity.

**Comment by clason** [comment]
> Added tests, but they fail if I move the last three highlighting queries to the top as suggested https://github.com/nvim-treesitter/nvim-treesitter/pull/5021#discussion_r1242453572

Yes, you need to use _our_ test suite (which unfortunately is not trivial to run locally); `tree-sitter test` will give wrong results due to the different query ordering strategies.

File: tests/query/highlights/wing/class.w [hunk 8]
```diff
@@ -0,0 +1,19 @@
+bring cloud;
+// <- @keyword
```
Comment (clason) [inline tests/query/highlights/wing/class.w:2]:
```suggestion
// <- keyword
```
omit the leading `@`

Comment (clason) [reply tests/query/highlights/wing/class.w]: everywhere; but the test still fails for some reason. @theHamsta ?

**Comment by clason** [comment]
Oh, and you probably also need to add the `.w` extension for `wing` files to
https://github.com/nvim-treesitter/nvim-treesitter/blob/bad9f5bf5b4af5d97a9cffdc957f1b23dc91d851/scripts/minimal_init.lua#L5-L15

File: queries/wing/highlights.scm [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  name: (identifier) @field)
+(class_definition 
+  name: (identifier) @type
+)
+(method_definition
+  name: (identifier) @method
+)
+(inflight_method_definition
+  name: (identifier) @method
+)
```
Comment (ObserverOfTime) [inline queries/wing/highlights.scm:18]:
```suggestion
  name: (identifier) @method)
```

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]
There are are more lines that should be reformatted. :)

**Review by ObserverOfTime — APPROVED** [review]

**Pull request merged**